poorwsgi/digest.py

```diff
@@ -0,0 +1,351 @@
+"""HTTP Authenticate Digest method.
+
+This file could be used as known ``htdigest`` tool.
+
+.. code ::sh
+
+    python3 -m poorwsgi.digest --help
+
+    # adding / updating user password
+    python3 -m poorwsgi.digest -c auth.digest 'Secret Zone' user
+"""
+from argparse import ArgumentParser
+from hashlib import md5, sha256
+from traceback import print_exc
+from getpass import getpass
+from os.path import exists
+from logging import getLogger
+from functools import wraps
+from collections import defaultdict
+
+import sys
+
+from poorwsgi import state
+from poorwsgi.response import HTTPException
+from poorwsgi.session import check_token
+
+log = getLogger("poorwsgi")  # pylint: disable=invalid-name
+
+
+def check_response(req, password):
+    """Check digest response value username/password in fact.
+
+    Return True if response value is right.
+    """
+
+    kwargs = req.authorization.copy()
+    kwargs['hash1'] = password
+
+    # session algorithm support
+    if req.app.auth_algorithm.endswith('-sess'):
+        hash1 = req.app.auth_hash(
+            "{hash1}:{nonce}:{cnonce}".format(**kwargs).encode()).hexdigest()
+        kwargs['hash1'] = hash1
+
+    kwargs['method'] = req.method
+
+    hash2 = req.app.auth_hash(
+        '{method}:{uri}'.format(**kwargs).encode()
+        ).hexdigest()
+    # auth-int support XXX: not tested
+    # md5_body = algorithm(req.data.encode()).hexdigest()
+    # hash2 = algorithm(
+    #     '{req.method}:{uri}:{md5_body}'
+    #     ''.format(req=req, md5_body=md5_body).encode()
+    #     ).hexdigest()
+    kwargs['hash2'] = hash2
+
+    # auth qop
+    if req.app.auth_qop:
+        # TODO: check nc value on server side
+        response = req.app.auth_hash(
+            '{hash1}:{nonce}:{nc}:{cnonce}:{qop}:{hash2}'
+            ''.format(**kwargs).encode()).hexdigest()
+    else:
+        response = req.app.auth_hash(
+            '{hash1}:{nonce}:{hash2}'
+            ''.format(**kwargs).encode()).hexdigest()
+    return response == kwargs['response']
+
+
+def check_credentials(req, realm, username=None):
+    """Check Digest authorization credentials.
+
+    Return True if Authorization header is valid for realm and username
+    if is set.
```

Comment:
`if is set` ... it feels like something is missing here

Reply:
username is not required.